File: tests/cpp/test_vision_protocol.cpp
#include <cassert>
#include <cstdint>
#include <vector>

#include "../../esp32/vision_receiver/vision_protocol.h"


static std::vector<uint8_t> bytes(const char* hex) {
    std::vector<uint8_t> result;
    while (*hex) {
        char pair[3] = {hex[0], hex[1], 0};
        result.push_back(static_cast<uint8_t>(strtoul(pair, nullptr, 16)));
        hex += 2;
    }
    return result;
}


int main() {
    using namespace vision;

    Parser parser;
    Frame frame;
    auto heartbeat = bytes("a55a01010001000003e80000f3c7");
    bool ready = false;
    for (uint8_t byte : heartbeat) {
        ready = parser.feed(byte, frame);
    }
    assert(ready);
    assert(frame.type == MessageType::Heartbeat);
    assert(frame.sequence == 1);
    assert(frame.timestampMs == 1000);

    auto face = bytes(
        "a55a01020002000003f2000f0000000001f401f40000041a00004bc36c"
    );
    ready = false;
    for (uint8_t byte : face) {
        ready = parser.feed(byte, frame);
    }
    assert(ready);
    FaceObservation observation;
    assert(decodeFace(frame, observation));
    assert(observation.centerX == 0);
    assert(observation.centerY == 0);
    assert(observation.width == 500);
    assert(observation.height == 500);
    assert(observation.yawCentidegrees == 1050);
    assert(observation.confidence == 75);

    auto corrupt = heartbeat;
    corrupt.back() ^= 0xFF;
    ready = false;
    for (uint8_t byte : corrupt) {
        ready = parser.feed(byte, frame) || ready;
    }
    assert(!ready);
    for (uint8_t byte : heartbeat) {
        ready = parser.feed(byte, frame) || ready;
    }
    assert(ready);

    VisionFreshness freshness(500);
    assert(!freshness.available(1000));
    freshness.noteValidFrame(1000);
    assert(freshness.available(1500));
    assert(!freshness.available(1501));

    return 0;
}

File: tests/test_vision_protocol.py
import unittest

from src.transport.vision_protocol import (
    MSG_FACE,
    MSG_HEARTBEAT,
    StreamDecoder,
    crc16_ccitt,
    decode_frame,
    encode_frame,
    pack_face_observation,
    unpack_face_observation,
)


class VisionProtocolTest(unittest.TestCase):
    def test_crc_matches_ccitt_false_reference(self):
        self.assertEqual(crc16_ccitt(b"123456789"), 0x29B1)

    def test_frame_round_trip(self):
        encoded = encode_frame(MSG_HEARTBEAT, b"", sequence=42, timestamp_ms=1234)

        decoded = decode_frame(encoded)

        self.assertEqual(decoded["type"], MSG_HEARTBEAT)
        self.assertEqual(decoded["sequence"], 42)
        self.assertEqual(decoded["timestamp_ms"], 1234)
        self.assertEqual(decoded["payload"], b"")

    def test_face_observation_is_resolution_independent(self):
        payload = pack_face_observation(
            frame_width=1920,
            frame_height=1080,
            box=[480, 270, 960, 540],
            euler=[12.34, -5.67, 1.25],
            confidence=0.92,
        )

        observation = unpack_face_observation(payload)

        self.assertEqual(observation["center_x"], 0)
        self.assertEqual(observation["center_y"], 0)
        self.assertEqual(observation["width"], 500)
        self.assertEqual(observation["height"], 500)
        self.assertEqual(observation["pitch_cd"], 1234)
        self.assertEqual(observation["yaw_cd"], -567)
        self.assertEqual(observation["roll_cd"], 125)
        self.assertEqual(observation["confidence"], 92)

    def test_pack_canonical_face_observation(self):
        payload = pack_face_observation(
            {
                "center_x": 0,
                "center_y": -250,
                "width": 500,
                "height": 333,
                "pitch_cd": 1234,
                "yaw_cd": -567,
                "roll_cd": 125,
                "confidence": 92,
            }
        )

        observation = unpack_face_observation(payload)

        self.assertEqual(observation["center_x"], 0)
        self.assertEqual(observation["center_y"], -250)
        self.assertEqual(observation["width"], 500)
        self.assertEqual(observation["height"], 333)
        self.assertEqual(observation["pitch_cd"], 1234)
        self.assertEqual(observation["yaw_cd"], -567)
        self.assertEqual(observation["roll_cd"], 125)
        self.assertEqual(observation["confidence"], 92)

    def test_pack_canonical_face_observation_requires_all_fields(self):
        with self.assertRaises(ValueError):
            pack_face_observation({"center_x": 0})

    def test_stream_decoder_recovers_after_corrupt_frame(self):
        corrupt = bytearray(
            encode_frame(MSG_HEARTBEAT, b"", sequence=1, timestamp_ms=10)
        )
        corrupt[-1] ^= 0xFF
        valid = encode_frame(MSG_FACE, b"abc", sequence=2, timestamp_ms=20)
        decoder = StreamDecoder()

        frames = decoder.feed(b"noise" + bytes(corrupt) + valid)

        self.assertEqual(len(frames), 1)
        self.assertEqual(frames[0]["sequence"], 2)
        self.assertEqual(frames[0]["payload"], b"abc")

    def test_sequence_accepts_wraparound_value(self):
        encoded = encode_frame(
            MSG_HEARTBEAT, b"", sequence=0xFFFF, timestamp_ms=0xFFFFFFFF
        )

        decoded = decode_frame(encoded)

        self.assertEqual(decoded["sequence"], 0xFFFF)
        self.assertEqual(decoded["timestamp_ms"], 0xFFFFFFFF)


if __name__ == "__main__":
    unittest.main()
